File: src/Inverse Square Root Algorithms/fisra-quake-III.py
import math
import time
import struct


def quake3_inv_sqrt(x):
    xhalf = 0.5 * x
    i = struct.unpack("<I", struct.pack("<f", x))[
        0
    ]  # store floating-point bits in integer
    i = 0x5F3759DF - (i >> 1)  # initial guess
    y = struct.unpack("<f", struct.pack("<I", i))[0]
    return y


x = 2.0
start_time = time.time()
result = quake3_inv_sqrt(x)
end_time = time.time()

print(f"Quake III Inverse Square Root of {x}: {result}")
print(f"Execution time: {(end_time - start_time) * 1e6:.2f} microseconds")

File: src/Inverse Square Root Algorithms/fisra-quake-III.cpp
#include <cmath>
#include <iostream>
#include <chrono>

float quake3InvSqrt(float x) {
    float xhalf = 0.5f * x;
    int i = *(int*)&x; // store floating-point bits in integer
    i = 0x5f3759df - (i >> 1); // initial guess
    float y = *(float*)&i;
    return y;
}

int main() {
    float x = 2.0f;
    auto start = std::chrono::high_resolution_clock::now();
    float result = quake3InvSqrt(x);
    auto end = std::chrono::high_resolution_clock::now();

    std::cout << "Quake III Inverse Square Root of " << x << ": " << result << std::endl;
    std::cout << "Execution time: " << std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() << " microseconds" << std::endl;

    return 0;
}